tutorials/crunch-aggregate-sessions.md

```diff
@@ -0,0 +1,296 @@
+```
+
+Step 2. Process the data.
+a. Create a session key by combining the user ID and session ID.
+b. Group together all events with the same session key.
+c. For each group, create a Session record as an Avro Specific object. 
+
+```Java
+    PCollection<Session> sessions = events
+        .by(new GetSessionKey(), Avros.strings())
+        .groupByKey()
+        .parallelDo(new MakeSession(), Avros.specifics(Session.class));
+
+```
+
+Step 3. Append the derived sessions to the target dataset, defined using the Kite `CrunchDatasets.asTarget` method.
+
+```Java
+    getPipeline().write(sessions,
+        CrunchDatasets.asTarget("dataset:hive:sessions"),
+        Target.WriteMode.APPEND);
+    return run().succeeded() ? 0 : 1;
+  }
+```
+
+The `GetSessionKey` class combines the session ID with the user ID to create a unique session key.
+
+```Java
+  private static class GetSessionKey extends MapFn<StandardEvent, String> {
+    @Override
+    public String map(StandardEvent event) {
+      // Create a key by combining the session id and user id
+      return event.getSessionId() + event.getUserId();
+    }
+  }
+```
+
+The `MakeSession` class extends `DoFn` (Do Function), where you handle transformation of your data before it goes into the target dataset.
+
+```Java
+
+  private static class MakeSession
+      extends DoFn<Pair<String, Iterable<StandardEvent>>, Session> {
+
+```
+
+The `process` method iterates through a group of events that have the same session key.
+
+```Java
+    @Override
+    public void process(
+        Pair<String, Iterable<StandardEvent>> keyAndEvents,
+        Emitter<Session> emitter) {
+      final Iterator<StandardEvent> events = keyAndEvents.second().iterator();
+      if (!events.hasNext()) {
+        return;
+      }
+
+```
+
+Initialize the values needed to create a session object for this group.
+
+```Java
+      final StandardEvent firstEvent = events.next();
+      long startTime = firstEvent.getTimestamp();
+      long endTime = firstEvent.getTimestamp();
+      int numEvents = 1;
+```
+
+Inspect each event in this session group. Track the earliest timestamp (start time) and latest timestamp (end time). Keep a count of the events in this session group.
+
+```Java
+      while (events.hasNext()) {
+        final StandardEvent event = events.next();
+        
+```
+
+Reset the start time if the timestamp is earlier than the current start time value.
+
+```Java
+        startTime = Math.min(startTime, event.getTimestamp());
+```
+
+Reset the end time if the timestamp is later then the current end time value.
+
+```Java
+        endTime = Math.max(endTime, event.getTimestamp());
+```
+
+Keep a count of the events in this session group.
+
+```Java
+        numEvents += 1;
+      }
+```
+
+Create a session using the Crunch `emitter`. Use values from the first event in the group for fields that don't change. Crunch repeats the `emit` process for each group of events until all sessions are processed.
```

Comment:
The emitter isn't used to create the session. That's the job of the Session builder, which is created with `Session.newBuilder()`. The emitter is used to send the session to the next phase of processing.

Crunch doesn't repeat emit, it keeps running this method for each group of events. This method emits a single session for each group.